packages/compass-aggregations/src/modules/pipeline-builder/builder-helpers.ts

```diff
@@ -1,31 +1,30 @@
 import type { PipelineBuilderThunkAction, RootState } from '..';
 import type { PipelineBuilder } from './pipeline-builder';
 import { loadPreviewForStagesFrom } from './stage-editor';
+import { loadPreviewForPipeline } from './text-editor';
 
 export const updatePipelinePreview =
   (): PipelineBuilderThunkAction<void> =>
-  (dispatch, getState, { pipelineBuilder }) => {
-    // Stop unconditionally (we might not start again based on the autoPreview
-    // state and stages validity)
-    pipelineBuilder.stopPreview();
+    (dispatch, getState, { pipelineBuilder }) => {
+      // Stop unconditionally (we might not start again based on the autoPreview
+      // state and stages validity)
+      pipelineBuilder.stopPreview();
 
-    if (getState().pipelineBuilder.pipelineMode === 'builder-ui') {
-      dispatch(loadPreviewForStagesFrom(0));
-    } else {
-      // TODO: dispatch for text editor
-    }
-  };
+      if (getState().pipelineBuilder.pipelineMode === 'builder-ui') {
+        dispatch(loadPreviewForStagesFrom(0));
+      } else {
+        void dispatch(loadPreviewForPipeline());
+      }
+    };
 
 export function getStagesFromBuilderState(
   state: RootState,
   pipelineBuilder: PipelineBuilder
 ) {
-  if (state.pipelineBuilder.pipelineMode === 'builder-ui') {
-    return pipelineBuilder.stages.map((stage) => stage.toBSON());
-  } else {
-    // TODO
-    return [];
+  if (state.pipelineBuilder.pipelineMode === 'as-text') {
+    pipelineBuilder.sourceToStages();
```

Comment:
I am happy to chat through this a bit more if you want to, from my perspective it would be a bit cleaner if we don't change / commit builder state unless the view type is switched. For that reason I would suggest to use `getPipelineFromSource` here instead of committing source to stages and then returning stage operators from stages. It also means that you don't need to go through the source > stages > stage to bson flow, you will just immediately go from source to bson / parsed pipeline